src/misc.ts

```diff
@@ -104,27 +113,34 @@ export async function httpStream (
 
   if (protocol.match(/^https:/i)) {
     get           = https.get
-    options       = parsedUrl
     options.agent = https.globalAgent
   } else if (protocol.match(/^http:/i)) {
     get           = http.get
-    options       = parsedUrl
     options.agent = http.globalAgent
   } else {
     throw new Error('protocol unknown: ' + protocol)
   }
 
   options.headers = {
-    ...options.headers,
     ...headers,
   }
 
-  const res = await new Promise<http.IncomingMessage>((resolve, reject) => {
-    get(options, resolve)
-      .on('error', reject)
+  return new Promise<http.IncomingMessage>((resolve, reject) => {
+    let res: http.IncomingMessage | null = null
+    const req = get(parsedUrl, options, (response) => {
+      res = response
+      resolve(res)
+    })
+      .once('error', reject)
+      .setTimeout(HTTP_TIMEOUT, () => {
+        const e = new Error('Http request timeout!')
+        if (res) {
+          res.destroy(e)
```

Comment:
Can we just emit the `error` instead of destroying the `res`?

Reply:
No. If we just emit error, res and req won't close themselves, so the code will still get stuck, causing a memory leak.

Reply:
I don't think so.

If the `error` is dealt with correctly by the downstream, then the resource will be destroyed, or rubbish collected, without any memory leak.

If the `error` is not dealt with correctly, then I believe the system will have an uncaught error event that will crash the Node.js v18+ (? not sure about that, need to be confirmed)

Reply:
>Sets the socket to timeout after timeout milliseconds of inactivity on the socket. By default net.Socket do not have a timeout.
>
>When an idle timeout is triggered the socket will receive a ['timeout'](https://nodejs.org/docs/latest-v18.x/api/net.html#event-timeout) event but the connection will not be severed. The user must manually call [socket.end()](https://nodejs.org/docs/latest-v18.x/api/net.html#socketenddata-encoding-callback) or [socket.destroy()](https://nodejs.org/docs/latest-v18.x/api/net.html#socketdestroyerror) to end the connection.

- https://nodejs.org/docs/latest-v18.x/api/http.html#requestsettimeouttimeout-callback
- https://nodejs.org/docs/latest-v18.x/api/net.html#socketsettimeouttimeout-callback

Reply:
In addition, please refer to the processing logic of ENOTFOUND error inside nodejs. When the connection fails, there will be a close event after the error event, that is, destroy is called, not just the emit error event.
If we don't call destroy, but just emit error event, instance will not trigger close event.